### ISSUE TITLE
Category list isn't loading on Bordeaux beta

### ISSUE THREAD

**Issue by wazolab** [teritorio/vido#170]
This [URL](https://city-demo-bordeaux.beta.appcarto.teritorio.xyz/category/49284754) is failing with 502 Error.

**After Investigation**
`CategorySelector` component is too long to compute `menuItems` props as we have an Array of 400+ items.

We have to find a way to decrease computational work and / or defer it in order to make rendering process quicker.

**Comment by wazolab** [comment]
This issue comes from the following `while` loop :
```js
while (parentId) {
  if (!menuIndex[parentId])
    return undefined

  const name = menuIndex[parentId].menu_group?.name.fr
  if (name && menuIndex[parentId].parent_id)
    parents.push(name)

  parentId = menuIndex[parentId].parent_id
}
```

We are building each item's title by fetching parent's title recursively ... on a 400+ item array, it's too much.

I think we have to rethink the way we want to manipulate and construct (`MenuItem | ApiMenuCategory`) arrays. We are doing too much of transformation, sometimes we are duplicating the same transformation as well ...

As of today here is the way it works:

### `pages/category/[id].vue`
- Either retrieve `menuStore.menuItems` if exists or fetch `menuItems` from Menu API
- On `created()` we call `menuStore.fetchConfig()` :
  - Filter non-hidden items
  - Use `item.id` for key name
  - Group child items with their parent
  - Add filter entry for each item
  - Save `filters` and `menuItems` data in `menuStore`

### `components/PoisList/PoissList.vue`
- Get `menuStore:apiMenuCategory` array which is a map on `menuStore.menuItems` values without any undefined item.
- Pass `apiMenuCategory` to `CategorySelector` with `:menu-items` prop

### `components/PoisList/CategorySelector.vue`
- Pass computed `menuEntries` to `v-autocomplete`
- `menuEntries` computed is doing the following :
  - Filter non-hidden items on `menuItems` prop (:warning: already done in `menuStore.fetchConfig()`)
  - Create the same array with `item.id` as `key` as `menuStore.menuItems`, we could re-use directly `menuStore.menuItems`
  - Filter again non-hidden items and with category (maybe we could separte items with category earlier in the process)
  - Map each item in order to have `value` and `title` required fields by `v-autocomplete`. In order to build each item's title we are doing a recursive parent title fetching (which is the main issue on load)
  - Filter undefined entries (I don't know if it even possible to have undefined entry at this level ...)
  - Then we sort the array by item's title abc order
---

In my opinion `CategorySelector` should be served with an already shaped data, no data transformation should be done here.

But before going for a big refactoring we could try the following :
- 1 flat array with inital API Response normalized
- 1 array with the whole hierararchy only with id references

One main advantage to split into different arrays is about not triggering too much reactivity (we can shallowRef some state entries if we want) and re-rendering cycles.
Array manipulation is lighter when we only have ID references and not big objects

This way we could build any hierarchy (Menu or Items in autocomplete) by loop on our ID references array, then we get the data with a simple find by id on our normalized array.

For the item's title in auto-complete we could either use something like [this](https://www.techighness.com/post/javascript-find-key-path-in-deeply-nested-object-or-array/) which returns us the path of keys needed for item's title construction (not the best in my opinion as we are traveling the array backwards AND one additional time).
Or we could build the title within the same time we are building our hierrachical array of IDs (which make more sensee in my opinion as we are travelling the array one way)

Basically the whole idea is to Normalize (or flatten) State data from the store in order to simplify access and manipulations.

Resources:
- About State Normalization: https://redux.js.org/usage/structuring-reducers/normalizing-state-shapeStore#designing-a-normalized-state
- Lib we could use for data normalization: https://github.com/paularmstrong/normalizr/tree/master/docs
- Global resource on Pinia usage and best practices: https://borstch.com/blog/development/state-management-solutions-in-vuejs-3

**Comment by frodrigo** [comment]
```
while (parentId) {
  if (!menuIndex[parentId])
    return undefined

  const name = menuIndex[parentId].menu_group?.name.fr
  if (name && menuIndex[parentId].parent_id)
    parents.push(name)

  parentId = menuIndex[parentId].parent_id
}
```

The ontology is only 3 levels deep. The loop should iterate only twice for each entry. Event for 400 that not so much.

I found inconsistance data in the menu
```json
  {
    "id": 4215628124,
    "hidden": false,
    "category": {
      "icon": "teritorio teritorio-park",
      "name": {
        "en": "Park",
        "fr": "Parc"
      },
      "slug": "parc",
      "zoom": 14,
      "color_fill": "#8cc56f",
      "color_line": "#70b06a",
      "style_class": [
        "nature",
        "park",
        "park"
      ],
      "style_merge": true,
      "display_mode": "large",
      "search_indexed": true
    },
    "parent_id": 4215628124,
    "index_order": 2,
    "selected_by_default": false
  },
```

The menu entry define itself as it own parent, so it makes an infinite loop on the while.
